src/common/cockpitconf.c

```diff
@@ -0,0 +1,159 @@
+
+  if (!g_key_file_load_from_file (key_file, file_path, G_KEY_FILE_NONE, error))
+    {
+      g_key_file_free (key_file);
+      return FALSE;
+    }
+
+  groups = g_key_file_get_groups (key_file, NULL);
+  for (i = 0; groups[i] != NULL; i++)
+    {
+      section = g_hash_table_lookup (cockpit_conf, groups[i]);
+      if (section == NULL)
+        {
+          section = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+          g_hash_table_insert (cockpit_conf, g_strdup (groups[i]), section);
+        }
+
+      keys = g_key_file_get_keys (key_file, groups[i], NULL, &err);
+      g_return_val_if_fail (err == NULL, FALSE);
+
+      for (j = 0; keys[j] != NULL; j++)
+        {
+          value = g_key_file_get_value (key_file, groups[i], keys[j], &err);
+          g_return_val_if_fail (err == NULL, FALSE);
+          g_hash_table_insert (section, g_strdup (keys[j]), value);
+        }
+      g_strfreev (keys);
+    }
+  g_strfreev (groups);
+
+  g_key_file_free (key_file);
+  return TRUE;
+}
+
+static void
+cockpit_conf_init (void)
+{
+  const gchar *cockpit_conf_file = COCKPIT_CONFIG_FILE;
+  GError *error = NULL;
+
+  cockpit_conf = g_hash_table_new_full (g_str_hash, g_str_equal, g_free,
+                                        (GDestroyNotify)g_hash_table_unref);
+
+
+  if (!_load_key_file (cockpit_conf_file, &error))
+    {
+      if (!g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
+        {
+          g_message ("couldn't load configuration file: %s: %s",
+                     cockpit_conf_file, error->message);
+        }
+      g_clear_error (&error);
+    }
+
+  g_debug ("Loaded settings from: %s", cockpit_conf_file);
+}
+
+
+void
+cockpit_conf_cleanup (void)
+{
+  g_assert (cockpit_conf != NULL);
+  g_hash_table_destroy (cockpit_conf);
+  cockpit_conf = NULL;
+}
+
+
+static void
+ensure_cockpit_conf (void)
+{
+  if (cockpit_conf == NULL)
+    cockpit_conf_init ();
+}
+
+
+const gchar *
+cockpit_conf_string (const gchar *section,
+                     const gchar *field)
+{
+  GHashTable *sect = NULL;
+  ensure_cockpit_conf ();
+  sect = g_hash_table_lookup (cockpit_conf, section);
+  if (sect != NULL)
+    return g_hash_table_lookup (sect, field);
+
+  return NULL;
+}
+
+
+const gchar **
+cockpit_conf_strv (const gchar *section,
+                   const gchar *field,
+                   gchar delimiter)
+{
+  const gchar **value = NULL;
+  gchar **strv = NULL;
+  const gchar *string = NULL;
+
+  string = cockpit_conf_string (section, field);
+
+  if (string != NULL)
+    {
+      strv = g_strsplit (g_strstrip (g_strdup (string)),
+                         &delimiter, -1);
+      if (strv)
+        {
+          guint n = g_strv_length (strv);
+          guint i;
+          value = g_new (const gchar *, n + 1);
+
+          for (i = 0; i < n; i++)
+            value[i] = g_strdup (strv[i]);
+
+          value[i] = NULL;
+          g_strfreev (strv);
+        }
+    }
+
+  return value;
```

Comment:
This needs to be owned by something or it will leak. The idea is that the caller doesn't need to care.